[Security Solution] Adds UI support for filtering by rule source customization

dplumlee opened the pull request:
## Summary

Addresses https://github.com/elastic/kibana/issues/180169

> [!NOTE]  
> Feature is behind the `prebuiltRulesCustomizationEnabled` feature flag.

Adds a filter for prebuilt rules in the Update rules table for "Modified" and "Unmodified" rules. Also adds a badge column in the Rules table to display whether a prebuilt rule has been customized or not.

Also switches the "Customized Elastic rule" badge on the rule details page to align with the updated language of "_Modified_ Elastic rule"

### Screenshots

#### Modified badge in Rules table
![Screenshot 2024-11-05 at 3 05 56 PM](https://github.com/user-attachments/assets/1f3313bb-7171-42b5-99b0-b9fb296fefd3)

#### Modification filter dropdown on Rule update page
<img width="1479" alt="Screenshot 2024-10-24 at 11 46 26 AM" src="https://github.com/user-attachments/assets/82715abe-6ff6-4ba6-97b3-6fab9f42069e">

#### New "customized rule" badge language on Rule details page
![Screenshot 2024-11-05 at 3 14 58 PM](https://github.com/user-attachments/assets/4e22ba3a-e13f-4cf1-88c0-6b5b0b2c258a)


### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))



### For maintainers

- [ ] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#_add_your_labels)
- [ ] This will appear in the **Release Notes** and follow the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)



<!--ONMERGE {"backportTargets":["8.x"]} ONMERGE-->

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/244851)
* Commit: 18e885b296398b36a393dd2d6bbcbaac0eedd1a4

### Failed CI Steps
* [Jest Tests #12](https://buildkite.com/elastic/kibana-pull-request/builds/244851#0192b603-43c6-45d9-92a1-4c4c38c24f5c)
* [test/functional/apps/discover/group3/config.ts](https://buildkite.com/elastic/kibana-pull-request/builds/244851#0192b603-43d0-4eaa-9a94-9f15501786b3)
* [test/functional/apps/discover/group3/config.ts](https://buildkite.com/elastic/kibana-pull-request/builds/244851#0192b627-be0f-4401-9666-a452a2f6bded)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/244851#0192b603-43d0-4eaa-9a94-9f15501786b3) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/244851/jobs/0192b603-43d0-4eaa-9a94-9f15501786b3/artifacts/0192b627-94c9-47f9-b493-4a0f47cd9d41) test/functional/apps/discover/group3/config.ts / discover/group3 discover lens vis should be able to load a saved search with custom vis, edit vis and revert changes
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/244851#0192b627-be0f-4401-9666-a452a2f6bded) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/244851/jobs/0192b627-be0f-4401-9666-a452a2f6bded/artifacts/0192b650-569f-45eb-a740-4bbaba99705d) test/functional/apps/discover/group3/config.ts / discover/group3 discover lens vis should be able to load a saved search with custom vis, edit vis and revert changes
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/244851#0192b603-43c6-45d9-92a1-4c4c38c24f5c) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/244851/jobs/0192b603-43c6-45d9-92a1-4c4c38c24f5c/artifacts/0192b626-3f60-4783-b138-b37feb1fe107) Jest Tests #<span></span>12 / ShowMoreButton calls onShowMoreClick on button click

cc @dplumlee

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/244851","number":244851,"commit":"18e885b296398b36a393dd2d6bbcbaac0eedd1a4"}],"number":244851}
buildkite-pr-comment-->

elasticmachine commented:
Pinging @elastic/security-detections-response (Team:Detections and Resp)

elasticmachine commented:
Pinging @elastic/security-solution (Team: SecuritySolution)

elasticmachine commented:
Pinging @elastic/security-detection-rule-management (Team:Detection Rule Management)

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/245249)
* Commit: 6300a5272a3678840c71fea6814fdb7c331792ab

### Failed CI Steps
* [FTR Configs #29](https://buildkite.com/elastic/kibana-pull-request/builds/245249#0192ba94-1f43-41b4-a6f7-b9930b676d38)
* [FTR Configs #31](https://buildkite.com/elastic/kibana-pull-request/builds/245249#0192ba94-1f46-4084-8169-0b3e4019bd6b)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/48fe299d5bcd230d63ce2dd8e89c62307dd0ee82) | [after](https://github.com/elastic/kibana/commit/6300a5272a3678840c71fea6814fdb7c331792ab) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 6041 | 6042 | +1 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/48fe299d5bcd230d63ce2dd8e89c62307dd0ee82) | [after](https://github.com/elastic/kibana/commit/6300a5272a3678840c71fea6814fdb7c331792ab) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 20.7MB | 20.7MB | +6.5KB |



### History
* :broken_heart: [Build #244851](https://buildkite.com/elastic/kibana-pull-request/builds/244851) failed 18e885b296398b36a393dd2d6bbcbaac0eedd1a4

cc @dplumlee

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/245249","number":245249,"commit":"6300a5272a3678840c71fea6814fdb7c331792ab"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/244851","number":244851,"commit":"18e885b296398b36a393dd2d6bbcbaac0eedd1a4"}],"number":245249}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/245688)
* Commit: 343e43df68f6e1a540ad219a265639a970d6b687
* [Cloud Deployment](https://kibana-pr-197340.kb.us-west2.gcp.elastic-cloud.com)

### History
* :yellow_heart: [Build #245249](https://buildkite.com/elastic/kibana-pull-request/builds/245249) was flaky 6300a5272a3678840c71fea6814fdb7c331792ab
* :broken_heart: [Build #244851](https://buildkite.com/elastic/kibana-pull-request/builds/244851) failed 18e885b296398b36a393dd2d6bbcbaac0eedd1a4

cc @dplumlee

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/245688","number":245688,"commit":"343e43df68f6e1a540ad219a265639a970d6b687"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/245249","number":245249,"commit":"6300a5272a3678840c71fea6814fdb7c331792ab"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/244851","number":244851,"commit":"18e885b296398b36a393dd2d6bbcbaac0eedd1a4"}],"number":245688}
buildkite-pr-comment-->

xcrzx commented:
I just started reviewing the PR and noticed an inconsistency in how we refer to customized rules.

**On the rule details page, we label them "Customized":**  
<img width="500px" src="https://github.com/user-attachments/assets/2fbf5586-695f-4d95-9b3c-4f665b479190" />

**On the rule table, we label them "Modified":**  
<img width="500px" src="https://github.com/user-attachments/assets/78c7a594-54e7-4eb9-bc8b-f5ab92157603" />

I think we should use consistent naming across different pages. I’m fine with either term, let’s just pick one and stick with it.

xcrzx commented:
@dplumlee, the ticked description also mentions:

> Change Integrations badge to just icon and number count as in the design below.

Should it also be implemented in this PR?

xcrzx commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/logic/types.ts:
```diff
@@ -202,3 +203,8 @@ export interface FindRulesReferencedByExceptionsProps {
 }
+
+export enum RuleSourceTypesEnum {
+  MODIFIED = 'MODIFIED',
+  UNMODIFIED = 'UNMODIFIED',
```
Nit: we use lowercase enum keys most of the time (when we don't need to map them on a string constant).

xcrzx commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/upgrade_prebuilt_rules_table/upgrade_prebuilt_rules_table_filters.tsx:
```diff
@@ -28,7 +31,11 @@ const UpgradePrebuiltRulesTableFiltersComponent = () => {
   } = useUpgradePrebuiltRulesTableContext();
 
-  const { tags: selectedTags } = filterOptions;
+  const isPrebuiltRulesCustomizationEnabled = useIsExperimentalFeatureEnabled(
+    'prebuiltRulesCustomizationEnabled'
+  );
```
This is repeated quite often on the frontend side. It might be worth extracting into its own hook, such as `useIsPrebuiltRulesCustomizationEnabled`.

xcrzx commented on x-pack/plugins/security_solution/public/detection_engine/rule_management/logic/types.ts:
```diff
@@ -202,3 +203,8 @@ export interface FindRulesReferencedByExceptionsProps {
   signal?: AbortSignal;
 }
+
+export enum RuleSourceTypesEnum {
```
We already have a rules source type that can be either `internal` or `external`. To avoid confusion with this enum, let's rename it to something like `RuleCustomizationEnum`.

xcrzx commented on x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/upgrade_prebuilt_rules_table/upgrade_rule_source_filter_popover.tsx:
```diff
@@ -0,0 +1,90 @@
+
+const RULE_SOURCE_POPOVER_WIDTH = 200;
+
+const RuleSourceFilterPopoverComponent = ({
```
`RuleSourceFilterPopoverComponent` -> `RuleCustomizationFilterPopoverComponent`. 

Please also update other occurrences of `RuleSource` to `RuleCustomization` within this file. This will future-proof the component, as we may introduce a table filter to differentiate external rules from internal ones. Distinguishing rule customization status from rule source type will help with that.

xcrzx reviewed (commented):
I tested the PR locally with the feature flag on and off and found no issues 👍 

I’m leaving a couple of suggestions to clarify naming. It would also be great to add a few happy path tests to cover the new functionality.

dplumlee commented:
> I think we should use consistent naming across different pages. I’m fine with either term, let’s just pick one and stick with it.
@xcrzx That's a good point, I would agree. I think the badge was added before we recently changed some of the language we were using, @ARWNightingale should we be using "Modified" or "Customized" to refer to the customized prebuilt rules in the UI

approksiu commented:
> Hey @approksiu, we've been using the word "customized" and "customization" all over the place - in tickets, PR descriptions, code, API contracts, and even the data we store in Elasticsearch in rule saved objects. Any good reason we should divert from it in favor of "modified"? If we do it, we will need to update the existing terminology everywhere, except the already closed issues. Renaming the `rule_source.is_customized` field in Elasticsearch won't be possible because we don't have a migration mechanism for rules.

We only need to change the badge on the Rule details page to "modified", no need to change "customized" naming  in code. The "Modified" badge is the interim solution till we have faceted filters and an additional indication of modified/customized fields on the rule page/flyout. cc @dplumlee @xcrzx @banderror

banderror commented:
The above sounds good to me, we discussed this with @approksiu offline too.

dplumlee commented:
@xcrzx Addressed your comments and added a couple cypress tests behind the feature flag. Tried to follow the pattern in [this PR](https://github.com/elastic/kibana/pull/196776) to try and align testing efforts

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/248625)
* Commit: e28e454eb0d4016cd2317700d6ac8e6647d0aa0e
* [Cloud Deployment](https://kibana-pr-197340.kb.us-west2.gcp.elastic-cloud.com)

### History
* :yellow_heart: [Build #245688](https://buildkite.com/elastic/kibana-pull-request/builds/245688) was flaky 343e43df68f6e1a540ad219a265639a970d6b687
* :yellow_heart: [Build #245249](https://buildkite.com/elastic/kibana-pull-request/builds/245249) was flaky 6300a5272a3678840c71fea6814fdb7c331792ab
* :broken_heart: [Build #244851](https://buildkite.com/elastic/kibana-pull-request/builds/244851) failed 18e885b296398b36a393dd2d6bbcbaac0eedd1a4

cc @dplumlee

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"running","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/248625","number":248625,"commit":"e28e454eb0d4016cd2317700d6ac8e6647d0aa0e"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/245688","number":245688,"commit":"343e43df68f6e1a540ad219a265639a970d6b687"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/245249","number":245249,"commit":"6300a5272a3678840c71fea6814fdb7c331792ab"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/244851","number":244851,"commit":"18e885b296398b36a393dd2d6bbcbaac0eedd1a4"}],"number":248625}
buildkite-pr-comment-->

xcrzx reviewed (approved):
Thanks for addressing my comments, @dplumlee 👍
Verified the changes locally, everything looks good!

dplumlee commented:
@elasticmachine merge upstream

dplumlee commented:
@elasticmachine merge upstream

kibanamachine commented:
Starting backport for target branches: 8.x

https://github.com/elastic/kibana/actions/runs/11710836929

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/248971)
* Commit: cad65bda37b18dde322ce78e81f34741a49e70b0

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/bc313f4b48680624a0f778eb02eb10158f604a15) | [after](https://github.com/elastic/kibana/commit/cad65bda37b18dde322ce78e81f34741a49e70b0) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 6139 | 6141 | +2 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/bc313f4b48680624a0f778eb02eb10158f604a15) | [after](https://github.com/elastic/kibana/commit/cad65bda37b18dde322ce78e81f34741a49e70b0) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 21.0MB | 21.0MB | +8.4KB |



### History
* :broken_heart: [Build #248916](https://buildkite.com/elastic/kibana-pull-request/builds/248916) failed acd5f59282c2ce09feac8a4d469d6cb2d341bc78
* :green_heart: [Build #248625](https://buildkite.com/elastic/kibana-pull-request/builds/248625) succeeded e28e454eb0d4016cd2317700d6ac8e6647d0aa0e
* :yellow_heart: [Build #245688](https://buildkite.com/elastic/kibana-pull-request/builds/245688) was flaky 343e43df68f6e1a540ad219a265639a970d6b687
* :yellow_heart: [Build #245249](https://buildkite.com/elastic/kibana-pull-request/builds/245249) was flaky 6300a5272a3678840c71fea6814fdb7c331792ab
* :broken_heart: [Build #244851](https://buildkite.com/elastic/kibana-pull-request/builds/244851) failed 18e885b296398b36a393dd2d6bbcbaac0eedd1a4

cc @dplumlee

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/248971","number":248971,"commit":"cad65bda37b18dde322ce78e81f34741a49e70b0"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/248916","number":248916,"commit":"acd5f59282c2ce09feac8a4d469d6cb2d341bc78"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/248625","number":248625,"commit":"e28e454eb0d4016cd2317700d6ac8e6647d0aa0e"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/245688","number":245688,"commit":"343e43df68f6e1a540ad219a265639a970d6b687"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/245249","number":245249,"commit":"6300a5272a3678840c71fea6814fdb7c331792ab"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/244851","number":244851,"commit":"18e885b296398b36a393dd2d6bbcbaac0eedd1a4"}],"number":248971}
buildkite-pr-comment-->

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/199217">](https://github.com/elastic/kibana/pull/199217)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 9.4.3 -->
